http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/AppComponent.scala

```diff
@@ -0,0 +1,165 @@
+import org.broadinstitute.dsde.workbench.leonardo.db.LeoProfile.{dummyDate, unmarshalDestroyedDate}
+
+import scala.concurrent.ExecutionContext
+
+final case class AppRecord(id: AppId,
+                           nodepoolId: NodepoolLeoId,
+                           appType: AppType,
+                           appName: AppName,
+                           status: AppStatus,
+                           samResourceId: AppSamResourceId,
+                           creator: WorkbenchEmail,
+                           createdDate: Instant,
+                           destroyedDate: Instant,
+                           dateAccessed: Instant,
+                           namespaceId: NamespaceId,
+                           diskId: Option[DiskId])
+
+class AppTable(tag: Tag) extends Table[AppRecord](tag, "APP") {
+  //unique (appName, destroyedDate)
+  def id = column[AppId]("id", O.PrimaryKey, O.AutoInc)
+  def nodepoolId = column[NodepoolLeoId]("nodepoolId")
+  def appType = column[AppType]("appType", O.Length(254))
+  def appName = column[AppName]("appName", O.Length(254))
+  def status = column[AppStatus]("status", O.Length(254))
+  def samResourceId = column[AppSamResourceId]("samResourceId", O.Length(254))
+  def creator = column[WorkbenchEmail]("creator", O.Length(254))
+  def createdDate = column[Instant]("createdDate", O.SqlType("TIMESTAMP(6)"))
+  def destroyedDate = column[Instant]("destroyedDate", O.SqlType("TIMESTAMP(6)"))
+  def dateAccessed = column[Instant]("dateAccessed", O.SqlType("TIMESTAMP(6)"))
+  def namespaceId = column[NamespaceId]("namespaceId", O.Length(254))
+  def diskId = column[Option[DiskId]]("diskId", O.Length(254))
+
+  def * =
+    (
+      id,
+      nodepoolId,
+      appType,
+      appName,
+      status,
+      samResourceId,
+      creator,
+      createdDate,
+      destroyedDate,
+      dateAccessed,
+      namespaceId,
+      diskId
+    ) <> (AppRecord.tupled, AppRecord.unapply)
+}
+
+object appQuery extends TableQuery(new AppTable(_)) {
+  def unmarshalApp(app: AppRecord,
+                   sevices: List[KubernetesService],
+                   labels: LabelMap,
+                   namespace: Namespace,
+                   disk: Option[PersistentDisk]): App =
+    App(
+      app.id,
+      app.nodepoolId,
+      app.appType,
+      app.appName,
+      app.status,
+      app.samResourceId,
+      AuditInfo(
+        app.creator,
+        app.createdDate,
+        unmarshalDestroyedDate(app.destroyedDate),
+        app.dateAccessed
+      ),
+      labels,
+      AppResources(
+        namespace,
+        disk,
+        sevices
+      ),
+      List()
+    )
+
+  def save(saveApp: SaveApp)(implicit ec: ExecutionContext): DBIO[App] = {
+    val namespaceName = saveApp.app.appResources.namespace.name
+    for {
+      nodepool <- nodepoolQuery
+        .getMinimalById(saveApp.app.nodepoolId)
+        .map(
+          _.getOrElse(
+            throw new SQLIntegrityConstraintViolationException(
```

Comment:
we do this instead of allowing the slick `.save` to handle the constraint violation because we want to use the `clusterId` associated with the `nodepoolId` to save the app's namespace, instead of allowing the caller to specify the clusterId for the namespace.

Reply:
I think you can do `.flatMap{x => x.fold(DBIO.failed(exception), DBIO.successful)}